File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.java
class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        Set<String> supplySet = new HashSet<>(Arrays.asList(supplies));
        Map<String, List<String>> recipeIngredients = new HashMap<>();
        Set<String> result = new HashSet<>();
        
        for (int i = 0; i < recipes.length; i++) {
            recipeIngredients.put(recipes[i], ingredients.get(i));
        }
        
        boolean foundNewRecipe = true;
        while (foundNewRecipe) {
            foundNewRecipe = false;
            for (String recipe : recipes) {
                if (!result.contains(recipe)) {
                    List<String> requiredIngredients = recipeIngredients.get(recipe);
                    if (supplySet.containsAll(requiredIngredients)) {
                        supplySet.add(recipe);
                        result.add(recipe);
                        foundNewRecipe = true;
                    }
                }
            }
        }
        return new ArrayList<>(result);
    }
}

File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.py
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        supply_set = set(supplies) 
        recipe_map = {recipe: set(ingredient) for recipe, ingredient in zip(recipes, ingredients)}  
        result = []
        queue = list(supply_set)
        can_create = set(queue)

        while queue:
            ingredient = queue.pop(0)

            for idx, recipe in enumerate(recipes):
                if recipe not in can_create and recipe not in result:
                    if recipe_map[recipe].issubset(can_create):
                        result.append(recipe)
                        queue.append(recipe)
                        can_create.add(recipe)

        return result
